### PR TITLE
Openapi documentation

### PR DESCRIPTION
**Pull request by ylavoie**
- Add infrastructure to produce documentation for OpenAPI interface from the code source.
- Fix a few typos in the api

### CODE REVIEW

File: openapi/extract_data_section.pl [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/perl
+use strict;
+use warnings;
+
+while ( my $file=shift(@ARGV)) {
+    my $yaml=$file;
+    $yaml =~ s/.+\///;
```
Comment (ehuelsmann) [inline openapi/extract_data_section.pl:7]: This looks like a nice task for [File::Spec->splitpath()](https://perldoc.perl.org/File::Spec#splitpath).

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm [hunk 2]
```diff
@@ -436,8 +436,13 @@ info:
   title: Management of industry codes configuration
   version: 0.0.1
 paths:
-  /contacts/sic/:
+  /contacts/sic:
+    description: Management of Contacts
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Routes/ERP/API/Contacts.pm:440]:
```suggestion
    description: Management of Standard Industry Codes (SICs)
```

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm [hunk 3]
```diff
@@ -436,8 +436,13 @@ info:
...
     get:
+      tags:
+        - Contacts
+      summary: Get a list of Contacts
+      operationId: getContacts
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Routes/ERP/API/Contacts.pm:445]:
```suggestion
      summary: Get a list of SICs
      operationId: getSICs
```

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm [hunk 4]
```diff
@@ -447,7 +452,19 @@ paths:
...
+          $ref: '#/components/responses/401'
+        403:
+          $ref: '#/components/responses/403'
+        404:
+          $ref: '#/components/responses/404'
     post:
+      tags:
+        - Contacts
+      summary: Put a list of Contacts
+      operationId: putContacts
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Routes/ERP/API/Contacts.pm:467]:
```suggestion
      summary: Add SIC entry
      operationId: postSIC
```

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm [hunk 5]
```diff
@@ -472,6 +497,10 @@ paths:
           $ref: '#/components/schemas/sic-code'
         style: simple
     get:
+      tags:
+        - Contacts
+      summary: Get a single Contact
+      operationId: getContactsById
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Routes/ERP/API/Contacts.pm:503]:
```suggestion
      summary: Get a single SIC
      operationId: getSICByCode
```

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm [hunk 6]
```diff
@@ -493,6 +522,10 @@ paths:
         404:
           $ref: '#/components/responses/404'
     put:
+      tags:
+        - Contacts
+      summary: Put a single Contact
+      operationId: putContactsById
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Routes/ERP/API/Contacts.pm:528]:
```suggestion
      summary: Update a single SIC
      operationId: putSICByCode
```

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm [hunk 7]
```diff
@@ -531,6 +564,10 @@ paths:
         428:
           $ref: '#/components/responses/428'
     delete:
+      tags:
+        - Contacts
+      summary: Delete a single Contact
+      operationId: deleteContactsById
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Routes/ERP/API/Contacts.pm:570]:
```suggestion
      summary: Delete a single SIC
      operationId: deleteSICByCode
```

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm [hunk 8]
```diff
@@ -561,7 +623,19 @@ paths:
...
+          $ref: '#/components/responses/401'
+        403:
+          $ref: '#/components/responses/403'
+        404:
+          $ref: '#/components/responses/404'
     post:
+      tags:
+        - Business types
+      summary: Put business types
+      operationId: putBusinessTypes
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Routes/ERP/API/Contacts.pm:638]:
```suggestion
      summary: Create a business type
      operationId: putBusinessTypes
```

File: openapi/extract_data_section.pl [hunk 9]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/perl
+use strict;
+use warnings;
+
```
Comment (ehuelsmann) [inline openapi/extract_data_section.pl:4]:
```suggestion

use File::Spec;
```

File: openapi/extract_data_section.pl (hunk 1 above)
Comment (ehuelsmann) [inline openapi/extract_data_section.pl:7]:
```suggestion
    my (undef, undef, $yaml) = File::Spec->splitpath( $file );
```

**Review by ehuelsmann — CHANGES_REQUESTED** [review]
Reviewed the descriptions; some need adjusting.

File: openapi/_LedgerSMB.yaml [hunk 10]
```diff
@@ -0,0 +1,12 @@
+openapi: 3.0.0
```
[the remark this answers is not in the document]
Comment (ylavoie) [reply openapi/_LedgerSMB.yaml:1]: Done

File: openapi/extract_data_section.pl (hunk 1 above)
Comment (ylavoie) [reply openapi/extract_data_section.pl]: Rewritten

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm (hunk 2 above)
Comment (ylavoie) [reply lib/LedgerSMB/Routes/ERP/API/Contacts.pm]: Done

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm (hunk 3 above)
Comment (ylavoie) [reply lib/LedgerSMB/Routes/ERP/API/Contacts.pm]: Done

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm (hunk 4 above)
Comment (ylavoie) [reply lib/LedgerSMB/Routes/ERP/API/Contacts.pm]: Done

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm (hunk 5 above)
Comment (ylavoie) [reply lib/LedgerSMB/Routes/ERP/API/Contacts.pm]: Done

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm (hunk 6 above)
Comment (ylavoie) [reply lib/LedgerSMB/Routes/ERP/API/Contacts.pm]: Done

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm (hunk 7 above)
Comment (ylavoie) [reply lib/LedgerSMB/Routes/ERP/API/Contacts.pm]: Done

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm (hunk 8 above)
Comment (ylavoie) [reply lib/LedgerSMB/Routes/ERP/API/Contacts.pm]: Create business types.

File: openapi/extract_data_section.pl (hunk 9 above)
Comment (ylavoie) [reply openapi/extract_data_section.pl:4]: Rewritten

File: openapi/extract_data_section.pl (hunk 1 above)
Comment (ylavoie) [reply openapi/extract_data_section.pl]: Rewritten

File: lib/LedgerSMB/Routes/ERP/API.pm [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+use warnings;
+
+
+##############################################################################
+#
+#
+#     LANGUAGES
+#
+#
+#############################################################################
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Routes/ERP/API.pm:31]:
```suggestion
```

File: lib/LedgerSMB/Routes/ERP/API/Products.pm [hunk 12]
```diff
@@ -555,7 +617,19 @@ paths:
...
+        401:
+          $ref: '#/components/responses/401'
+        403:
+          $ref: '#/components/responses/403'
+        404:
+          $ref: '#/components/responses/404'
     post:
+      tags:
+        - Warehouses
+      summary: Put a list of warehouses
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Routes/ERP/API/Products.pm:631]:
```suggestion
      summary: Create a warehouse
```

File: lib/LedgerSMB/Routes/ERP/API/Products.pm [hunk 13]
```diff
@@ -601,6 +688,10 @@ paths:
         404:
           $ref: '#/components/responses/404'
     put:
+      tags:
+        - Warehouses
+      summary: Put a single warehouse
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Routes/ERP/API/Products.pm:693]:
```suggestion
      summary: Update a single warehouse
```

File: lib/LedgerSMB/Routes/ERP/API/Products.pm (hunk 12 above)
Comment (ylavoie) [reply lib/LedgerSMB/Routes/ERP/API/Products.pm:631]: We're not putting a single one but a list

Comment (ehuelsmann) [reply lib/LedgerSMB/Routes/ERP/API/Products.pm:631]: I'm not sure what you're saying, but usually a "POST" to the parent resource creates a single new child resource. That's what's happening here too. Just check the schema of the request body: `#/components/schemas/NewWarehouse` (which is a single object, not a list).

File: lib/LedgerSMB/Routes/ERP/API/Contacts.pm (hunk 8 above)
Comment (ylavoie) [reply lib/LedgerSMB/Routes/ERP/API/Contacts.pm]: Fixed

File: lib/LedgerSMB/Routes/ERP/API/Products.pm (hunk 12 above)
Comment (ylavoie) [reply lib/LedgerSMB/Routes/ERP/API/Products.pm]: Fixed

**Review by ehuelsmann — APPROVED** [review]

**Pull request merged**